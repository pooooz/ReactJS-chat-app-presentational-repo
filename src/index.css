body {
  margin: 0;
  font-family: 'Press Start 2P', cursive;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
    background-color: black;
}

#root {
    display: flex;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}
