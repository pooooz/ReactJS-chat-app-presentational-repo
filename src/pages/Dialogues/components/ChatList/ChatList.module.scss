$text-color: white;

.chatList {
  font-family: 'Press Start 2P', cursive;
  min-height: 929px;
  min-width: 200px;
  max-width: 360px;
  margin: 0;
  padding: 10px 10px 0 10px;
  ul {
    list-style: none;
  }
  background-color: rgba(46, 66, 7, 0.1);
}

.chatItem {
  margin: 10px 0 0 0;
  display: flex;
  justify-content: space-between;
}

.chatHeader {
  color: white;
  text-align: center;
}

.link {
  color: $text-color;
}

.form {
  display: flex;
  justify-content: space-between;
}

.input {
  max-width: 200px;
  font-family: 'Press Start 2P', cursive;
  color: $text-color;
  background-color: transparent;
  border: none;
  outline: none;
}

.inputWrap:before {
  content: '>';
  color: $text-color
}

.button {
  font-family: 'Press Start 2P', cursive;
  color: $text-color;
  background: transparent;
  border: none;
  cursor: pointer;
}

.deleteButton {
  font-family: 'Press Start 2P', cursive;
  color: #6b0404;
  background: transparent;
  cursor: pointer;
  border: none;
}