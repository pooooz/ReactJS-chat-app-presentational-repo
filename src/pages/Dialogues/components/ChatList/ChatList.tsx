import React, { FC, useState } from 'react';
import { NavLink } from 'react-router-dom';

import { useDispatch, useSelector } from 'react-redux';
import { selectChatList } from 'src/store/dialogues/selectors';
import { addChat, deleteChat } from 'src/store/dialogues/actions';

import styles from './ChatList.module.scss';

export const ChatList: FC = () => {
  const [name, setName] = useState('');
  const chatList = useSelector(
    selectChatList,
    (prev, next) => prev.length === next.length
  );
  const dispatch = useDispatch();

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (name) {
      dispatch(addChat(name));
      setName('');
    }
  };

  return (
    <aside className={styles.chatList}>
      <div className={styles.chatHeader}>Chats</div>
      <ul>
        {chatList.map((chat) => (
          <li className={styles.chatItem} key={chat.id}>
            <NavLink to={`/chats/${chat.name}`} className={styles.link}>
              {chat.name}
            </NavLink>
            <button
              className={styles.deleteButton}
              onClick={() => dispatch(deleteChat(chat.name))}
            >
              X
            </button>
          </li>
        ))}
      </ul>
      <form onSubmit={handleSubmit} className={styles.form}>
        <div className={styles.inputWrap}>
          <input
            type="text"
            className={styles.input}
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
        </div>
        <button className={styles.button} type="submit">
          Add chat
        </button>
      </form>
    </aside>
  );
};
