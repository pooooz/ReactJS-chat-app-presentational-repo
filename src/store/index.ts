import { combineReducers } from 'redux';
import { configureStore } from '@reduxjs/toolkit';
import storage from 'redux-persist/lib/storage';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
  REHYDRATE,
  persistReducer,
  persistStore,
} from 'redux-persist';

import { profileReducer } from 'src/store/profile/slice';
import { dialoguesReducer } from 'src/store/dialogues/slice';

const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['profile'],
};

export type StoreState = ReturnType<typeof rootReducer>;
const rootReducer = combineReducers({
  profile: profileReducer,
  dialogues: dialoguesReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  devTools: process.env.NODE_ENV !== 'production',
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
  reducer: persistedReducer,
});

export const persistor = persistStore(store);
