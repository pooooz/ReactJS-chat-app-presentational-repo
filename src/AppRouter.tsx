import React, { FC } from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';

import { Sidebar } from './components/Sidebar/Sidebar';
import { Dialogue } from './pages/Dialogues/Dialogue';
import { ChatList } from './pages/Dialogues/components/ChatList/ChatList';
import { Profile } from 'src/pages/Profile/Profile';

export const AppRouter: FC = () => {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Sidebar />}>
          <Route path="profile" element={<Profile />} />
          <Route path="chats">
            <Route index element={<ChatList />} />
            <Route path=":chatId" element={<Dialogue />} />
          </Route>
        </Route>

        <Route path="*" element={<h2 style={{ color: '#00BFFF' }}>404</h2>} />
      </Routes>
    </BrowserRouter>
  );
};
