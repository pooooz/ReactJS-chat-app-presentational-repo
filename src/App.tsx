import React, { FC, useState } from 'react';
import { defaultContext, ThemeContext } from 'src/utils/ThemeContext';
import { Provider } from 'react-redux';

import { AppRouter } from './components/AppRouter';
import { store } from 'src/store';

export const App: FC = () => {
  const [theme, setTheme] = useState(defaultContext.theme);

  const toggleTheme = () => {
    setTheme(theme === 'light' ? 'dark' : 'light');
  };

  return (
    <Provider store={store}>
      <ThemeContext.Provider value={{ theme, toggleTheme }}>
        <AppRouter />
      </ThemeContext.Provider>
    </Provider>
  );
};
