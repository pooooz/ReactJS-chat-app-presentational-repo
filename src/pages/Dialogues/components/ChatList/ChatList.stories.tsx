import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import { ChatList } from './ChatList';
import { store } from 'src/store';

export default {
  title: 'pages/Dialogues/components/ChatList',
  component: ChatList,
} as ComponentMeta<typeof ChatList>;

export const Default: ComponentStory<typeof ChatList> = () => (
  <Provider store={store}>
    <MemoryRouter>
      <ChatList />
    </MemoryRouter>
  </Provider>
);
