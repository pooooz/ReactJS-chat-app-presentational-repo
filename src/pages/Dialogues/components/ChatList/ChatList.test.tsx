import React from 'react';
import '@testing-library/jest-dom';
import '@testing-library/user-event';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import { store } from 'src/store';
import { ChatList } from './ChatList';

describe('ChatList test', () => {
  it('Render component', () => {
    const history = createMemoryHistory();
    render(
      <Provider store={store}>
        <Router location="/chats" navigator={history}>
          <ChatList />
        </Router>
      </Provider>
    );
  });

  it('Adding a chat', () => {
    const history = createMemoryHistory();
    render(
      <Provider store={store}>
        <Router location="/chats" navigator={history}>
          <ChatList />
        </Router>
      </Provider>
    );
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'initial' },
    });
    fireEvent.click(screen.getAllByRole('button', { name: /Add chat/ })[0]);
    expect(screen.getByText(/initial/)).toBeInTheDocument();
  });

  it('Deleting a chat', () => {
    const history = createMemoryHistory();
    render(
      <Provider store={store}>
        <Router location="/chats" navigator={history}>
          <ChatList />
        </Router>
      </Provider>
    );
    fireEvent.click(screen.getAllByRole('button', { name: /X/ })[0]);
    expect(() => screen.getByText(/default/)).toThrow();
  });

  it('Adding an empty chat', () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store}>
        <Router location="/chats" navigator={history}>
          <ChatList />
        </Router>
      </Provider>
    );
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: '' },
    });
    fireEvent.click(screen.getByRole('button', { name: /Add chat/ }));
    expect(asFragment()).toMatchSnapshot();
  });
});
