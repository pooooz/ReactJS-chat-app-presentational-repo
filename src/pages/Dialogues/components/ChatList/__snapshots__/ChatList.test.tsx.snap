// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatList test Adding an empty chat 1`] = `
<DocumentFragment>
  <aside
    class="chatList"
  >
    <div
      class="chatHeader"
    >
      Chats
    </div>
    <ul>
      <li
        class="chatItem"
      >
        <a
          class="link"
          href="/chats/initial"
        >
          initial
        </a>
        <button
          class="deleteButton"
        >
          X
        </button>
      </li>
    </ul>
    <form
      class="form"
    >
      <div
        class="inputWrap"
      >
        <input
          class="input"
          type="text"
          value=""
        />
      </div>
      <button
        class="button"
        type="submit"
      >
        Add chat
      </button>
    </form>
  </aside>
</DocumentFragment>
`;
